File: Algorithm/gray_code.py
class Solution:
    def grayCode(self,n):
        lst=[]
        for x in range(1<<n):
            lst.append((x>>1)^x)
        print len(lst)
        return lst
    def grayCode_1(self,n):
        if n==0:
            return [0]
        res=[0,1]
        for x in range(2,n+1):
            t=1<<x-1
            res+=[t+j for j in res[::-1]]
            print res
        return res
            


if __name__=='__main__':
    sol=Solution()
    print sol.grayCode(4)
    print sol.grayCode_1(4)

File: Algorithm/CPP/gray_code.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <algorithm>
# include<time.h>
using namespace std;

class Solution {
	public:
		vector<int> grayCode(int n)
		{
			vector<int> res(1, 0);
			for (auto i = 1; i < (1 << n); ++i)
			{
				res.push_back(res[i - 1] ^ (i&(-i)));  // find the rightmost bit and xor with the previous number
			}
			return res;
		}
};
int  main()
{
	Solution sol;




	for (auto num1 : sol.grayCode(4)) cout << num1 << " ";
	cout << endl;
	cin.get();
	return 0;
}
